### PR TITLE
feat: handle temporal columns in group bys

### PR DESCRIPTION
**Pull request by betodealmeida**
### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Query before was invalid:

```sql
SELECT FLOOR("__time" TO DAY) AS "__time",
       COUNT(*) AS "count"
FROM "druid"."wikipedia"
GROUP BY FLOOR("__time" TO DAY)
ORDER BY "count" DESC
LIMIT 100;

SELECT FLOOR("__time" TO DAY) AS "__timestamp",
       FLOOR("__time" TO DAY) AS "__time",
       COUNT(*) AS "count"
FROM "druid"."wikipedia"
WHERE FLOOR("__time" TO DAY) = '2016-06-27T00:00:00.000Z' -- THIS FAILS
GROUP BY FLOOR("__time" TO DAY),
         FLOOR("__time" TO DAY)
ORDER BY "count" DESC
LIMIT 10000;
```

Query after:

```sql
SELECT FLOOR("__time" TO DAY) AS "__time",
       COUNT(*) AS "count"
FROM "druid"."wikipedia"
GROUP BY FLOOR("__time" TO DAY)
ORDER BY "count" DESC
LIMIT 100;

SELECT FLOOR("__time" TO DAY) AS "__timestamp",
       FLOOR("__time" TO DAY) AS "__time",
       COUNT(*) AS "count"
FROM "druid"."wikipedia"
WHERE FLOOR("__time" TO DAY) = TIME_PARSE('2016-06-27T00:00:00+00:00') -- THIS IS CORRECT
GROUP BY FLOOR("__time" TO DAY),
         FLOOR("__time" TO DAY)
ORDER BY "count" DESC
LIMIT 10000;
```

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/db_engine_specs/druid.py [hunk 1]
```diff
@@ -100,3 +100,17 @@ def convert_dttm(cls, target_type: str, dttm: datetime) -> Optional[str]:
...
+        Convert from number of seconds since the epoch to a timestamp.
+        """
+        return "MILLIS_TO_TIMESTAMP({col} * 1000)"
+
+    @classmethod
+    def epoch_ms_to_dttm(cls) -> str:
+        """
+        Convert from number of milliseconds since the epoch to a timestamp.
+        """
+        return "MILLIS_TO_TIMESTAMP({col})"
```
Comment (villebro) [inline superset/db_engine_specs/druid.py:116]: Oh wow, I didn't know these weren't defined yet for Druid!

**Review by villebro — APPROVED** [review]
LGTM - just needs a rebase

File: superset/db_engine_specs/druid.py (hunk 1 above)
Comment (betodealmeida) [reply superset/db_engine_specs/druid.py:116]: Yeah, I implemented `epoch_to_dttm` while working on a solution, and decided to leave it there even if I wasn't using it, and add `epoch_ms_to_dttm` for completeness.

**Comment by john-bodley** [comment]
@betodealmeida and @villebro I don't believe this logic works for all cases. You could have a temporal field which is encoded as a string per the Python date format which shouldn't be converted to a `TIMESTAMP`. 

Isn't the correct logic to use the type of the column rather than assuming it's a `TIMESTAMP`? Furthermore `self.db_engine_spec.convert_dttm("TIMESTAMP", dttm)` could return `None` (even for the `TIMESTAMP` target type). I was thinking the logic should be of the form,

```python
if column_map[dimension].is_temporal and and isinstance(value, str):
    if rv := self.db_engine_spec.convert_dttm(
        column_map[dimension].type, 
        dateutil.parser.parse(value),
    ):
        value = text(rv)
```

**Comment by john-bodley** [comment]
@betodealmeida I have proposed a fix in https://github.com/apache/superset/pull/17312.

**Pull request merged**